typing/shape.ml

```diff
@@ -370,8 +370,7 @@ end) = struct
           | NAbs(clos_env, var, body, _body_nf) ->
               let arg = delay_reduce env arg in
               let env = bind { env with local_env = clos_env } var (Some arg) in
-              reduce env body
-              |> improve_uid t.uid
+              { (reduce env body) with uid = t.uid }
```

Comment:
@gasche @Octachron 

When changing the reduction for aliases I spent a lot of time thinking about this and why I felt I had to always "improve the uid" when reducing App and I think my current implementation is ~the~ a correct one:
- When we perform a functor application, we _always_ create a new module, if we bind it to a name, it has a new uid and the struct of this new module should have this uid associated with it. In some cases, the struct of this new module is a copy of the struct of another one, but we should still update the uid.
- Sometimes, the functor looks a lot like an alias: `module F(X) = M` or `module F(X) = X`, but I don't think these should be considered as aliases, they are more like copies. We could consider these as aliases for the shapes however, and keep track of the module that was "copied", but I don't really think it is necessary, and I don't see how I could do it.

I added a handfull of example with the mentioned cases, and I feel like the results are the one we should expect.
Here is your Proj / Struct example:

```ocaml
module Id(X : sig end) = X
module Struct = struct
  module L = List
end
[%%expect{|
{
 "Id"[module] -> Abs<.12>(X/369, X/369<.11>);
 }
module Id : functor (X : sig end) -> sig end
{
 "Struct"[module] ->
   {<.14>
    "L"[module] -> Alias(<.13>
                         CU Stdlib . "List"[module]);
    };
 }
module Struct : sig module L = List end
|}]

module App = Id(List) (* this should be a copy of X's structure with App's uid *)
module Proj = Struct.L
  (* this should have the Proj uid and be an alias to Struct.L *)
[%%expect{|
{
 "App"[module] -> (CU Stdlib . "List"[module])<.15>;
 }
module App : sig end
{
 "Proj"[module] -> Alias(<.16>
                         Alias(<.13>
                               CU Stdlib . "List"[module]));
 }
module Proj = Struct.L
|}]
```

Reply:
> If you think of them as copies, then the shape of a functor body (in typemod.ml, when typing Tmod_functor) should erase its uid, to avoid referring to an existing uid, right? (But then why would one handle functor bodies this way, and structure fields differently?)

I think the difference comes from the fact that applications create new structures (even if they are just a copy of an existing one) while projection do not (but they might be used to create aliases).

For coherence, I agree that "erasing" the `uid` from functors body would be a correct solution. A functor's body has no name, so it doesn't have an uid either.

However, I am not sure that I can do that distinction at the point in Typemod where these shapes are created. I will have a look tomorrow. If not, doing the distinction while reducing looks like a plausible workaround to that issue.

Without this change (always updating the application's results uids), some tests are wrong:

```ocaml
module M = struct end (* uid 0 *)
module F(X : sig end) = M
module App = F(List)
[%%expect{|
{
 "M"[module] -> {<.0>};
 }
module M : sig end
{
 "F"[module] -> Abs<.2>(X/278, {<.0>});
 }
module F : functor (X : sig end) -> sig end
{
 "App"[module] -> {<.0>};    (* <-- this should be 3, the uid of App's definition *)
 }
module App : sig end
|}]
```

Reply:
Let's go back to the basics here: we use uids to identify/represent bindings.

So, if I take the questions in @gasche's tests above, the answer is always: "we want the uid of App/Proj".

From that point of view, the `improve_uid` logic for applications and projections is just incorrect. And so are the calls to `Shape.set_uid_if_none` in typemod. We should always force the uid to be the one of the new binding.
This issue wasn't noticed at first because so far shapes had only been used for jumping to definition, and merlin tries his best to "traverse" aliases. But the incorrect behavior (although minor) can be observed on examples like :
```ocaml
module M = struct type t let x = 1 end
module N : sig type t end = M
module O = N
```
Trying to jump to the definition of N will move the cursor to the first line (the shape associated to `N` doesn't contain the uid of `N`, but reuse the one from the shape of `M`). This was not a major issue for locate (indeed the cursor move could even be considered "correct"), but is definitely incorrect when listing occurrences.

---

This brings up a new question: does `improve_uid` ever make sense?

It turns out that it does, but in a place where it's currently not called: when substituting a functor parameter for the concrete argument.
Consider the following example:
```ocaml
module F (X : sig end) = struct module M = X end
module N = F(struct end)
module O = N.M
```
Here `N.M` doesn't have an uid with the current implementation of `reduce` (feed this example to `ocamlc -dshape`) but it would if we called `improve_uid` after succesful substitutions.

Another way to handle this case without resorting to `improve_uid` would be to change our definition of uids: instead of representing bindings, they could represent "definitions". Then even unbound structures (like in the example above) would have a uid, and there'd be no problem when substituting. But that's a more profound change and definitely out of the scope of this PR (if we even want to implement it, which is unclear).

---

tl;dr: I believe this patch is correct, but doesn't go far enough.

Reply:
Personally I find the discrepancy between the Proj and the App case disturbing. I would like a parallel between the two cases -- or a profound reason why they need to be handled differently, not just "the testsuite output looks good like this". If @trefis explanation can lead us to a consistent implementation, I would like that.

Reply:
In the last commit [d14cf79](https://github.com/ocaml/ocaml/pull/12508/commits/d14cf79c502504820c4d80f87b991f35d95fa8e7) I propose a change in shape reduction inspired by Thomas' analysis that uniform the app and proj cases. uid "improvement" is now done on the `var` case. This fixes a few more issues illustrated in the testsuite (in [2ea20f9](https://github.com/ocaml/ocaml/pull/12508/commits/2ea20f97335ce6339c5a621719e0dec7c0babfa8)).